src/openfhe.jl

```diff
@@ -74,28 +70,124 @@ end
 
 """
     init_rotation!(context::SecureContext{<:OpenFHEBackend}, private_key::PrivateKey,
-                   shifts)
+                   shape::NTuple{N, Integer}, rotation_index...)
 
-Generate rotation keys for use with `OpenFHE.EvalRotate` using the `private_key` and for the
-rotation indexes in `shifts`. The keys are stored in the given `context`.
-Positive shift defines rotation to the right, e.g. a rotation with a shift 1:
-[1, 2, 3, 4] -> [4, 1, 2, 3].
-Negative shift defines rotation to the left, e.g. a rotation with a shift -1:
-[1, 2, 3, 4] -> [2, 3, 4, 1].
-
-Note: Here, indexes stored in `shifts` have reversed sign compared to rotation indexes used in
-OpenFHE.
+Generate all required rotation keys for use with `circshift` for the
+rotation indexes in `rotation_index` using the `private_key`. The keys are stored in the given `context`.
 
 See also: [`SecureContext`](@ref), [`OpenFHEBackend`](@ref), [`PrivateKey`](@ref)
 """
 function init_rotation!(context::SecureContext{<:OpenFHEBackend}, private_key::PrivateKey,
-                        shifts)
+                        shape::NTuple{N, Integer}, rotation_index...) where N
     cc = get_crypto_context(context)
-    OpenFHE.EvalRotateKeyGen(cc, private_key.private_key, -shifts)
+    # Get indexes for precompilation
+    indexes = get_index(context, shape, rotation_index)
+    # All indexes correspond to Base.circshift, but to use with OpenFHE, all indexes
+    # have to be negated.
+    OpenFHE.EvalRotateKeyGen(cc, private_key.private_key, -unique(indexes))
 
     nothing
 end
 
+function get_index(context::SecureContext{<:OpenFHEBackend}, shape::Tuple{Integer},
+                   rotation_index)
```

Comment:
`get_index` is very close to `getindex` but does something completely different. Also, it is customary in Julia to _not_ add types in a function signature unless really needed for dispatch. Thus I suggest to change this interface to something like the following:
```suggestion
function compute_rotation_indices(context, shape, shifts)
```

Reply:
I have renamed the function, but the type of `shape` is still required. the reason is that there are two `compute_rotation_indices` functions, this one for 1D, that precomputes OpenFHE rotation indices for 1D shift, and below for nD. nD `circshift` utilizes a 1D version, so `compute_rotation_indices` for nD actually computes required 1D shifts and passes them to `compute_rotation_indices` 1D to get OpenFHE rotation indicies. That is why the type of shape is relevant to differentiate between 1D and nD. I also added explanation comments to the code.

Reply:
If these are only internally used functions and there is no reason for them to have the same name, why don't you call them `compute_rotation_indices_1d` and `compute_rotation_indices_nd`? That would make it immediately clear why where are two versions and which one is called where, without having to reason about the types.

Reply:
Good idea! I've done this and could even shorten a bit the 1d version.

I have also added .typos.toml to ignore "nd"